src/ML/src/velMANNAutoregressiveInputBuilder.cpp

```diff
@@ -0,0 +1,371 @@
+    if (m_pimpl->fsm == Impl::FSM::Idle)
+    {
+        log()->error("{} Please initialize the class before calling advance.", logPrefix);
+        return false;
+    }
+
+    // first of all we invalidate the output
+    m_pimpl->fsm = Impl::FSM::OutputInvalid;
+
+    // evaluate the control point and the x coordinate of the final point of the Bezier curve
+    m_pimpl->curve.controlPoint[1] = 0;
+
+    // if the robot is walking forward (motion direction positive)
+    if (m_pimpl->input.motionDirection[0] >= 0)
+    {
+        m_pimpl->curve.controlPoint[0]
+            = m_pimpl->ellipsoidScalingFactor
+              * std::min(m_pimpl->ellipsoidForwardAxis, m_pimpl->input.motionDirection[0]);
+        m_pimpl->curve.finalPoint[0] = m_pimpl->input.motionDirection[0]
+                                       * m_pimpl->ellipsoidScalingFactor
+                                       * m_pimpl->ellipsoidForwardAxis;
+    }
+    // if the robot is walking backward (forward direction negative)
+    else
+    {
+        m_pimpl->curve.controlPoint[0]
+            = m_pimpl->ellipsoidScalingFactor
+              * std::max(-m_pimpl->ellipsoidBackwardAxis, m_pimpl->input.motionDirection[0]);
+        m_pimpl->curve.finalPoint[0] = m_pimpl->input.motionDirection[0]
+                                       * m_pimpl->ellipsoidScalingFactor
+                                       * m_pimpl->ellipsoidBackwardAxis;
+    }
+
+    m_pimpl->curve.finalPoint[1] = m_pimpl->input.motionDirection[1]
+                                   * m_pimpl->ellipsoidScalingFactor * m_pimpl->ellipsoidSideAxis;
+
+    // compute the future trajectory of the robot base
+    m_pimpl->output.desiredFutureBaseTrajectory
+        = m_pimpl->curve.curve.evaluateCurve(m_pimpl->curve.initialPoint,
+                                             m_pimpl->curve.controlPoint,
+                                             m_pimpl->curve.finalPoint);
+
+    // evaluate the future base velocity as numerical differentiation of the base position
+    //                 ┌                         ┐
+    // base_position = │ x0, x1, x2, x3, ..., xn │
+    //                 │ y0, y1, y2, y3, ..., yn │
+    //                 └                         ┘
+    // the base velocity is computed using the following steps
+    //
+    // 1. numerically differentiate the base position
+    //                 ┌                                                         ┐
+    // base_velocity = │ x1 - x0, x2 - x1, x3 - x2, x4 - x3, ..., x{n-1} - xn, 0 │
+    //                 │ y1 - y0, y2 - y1, y3 - y2, y4 - y3, ..., y{n-1} - yn, 0 |
+    //                 └                                                         ┘
+    //
+    // 2. normalize the base velocity computed in step 1 and multiply it by
+    //    the base velocity norm.
+    //
+    // 3. copy the second to last element of the base_velocity computed in step 2 to
+    //    the last element
```

Comment:
Is the z missing?

Reply:
No, the input assumes ground projections for simplicity.